tests/UnitTests/CoreGraphics/CGDataConsumerTests.mm

```diff
@@ -0,0 +1,122 @@
+    EXPECT_EQ(50, written);
+    EXPECT_EQ(0, compareData[0]);
+    EXPECT_EQ(1, compareData[1]);
+    EXPECT_EQ(49, compareData[49]);
+    EXPECT_EQ(0, compareData[50]);
+
+    uint8_t* largeCompareData = (uint8_t*)calloc(100000, sizeof(uint8_t));
+    CGDataConsumerCallbacks largeCallbacks{ [](void* info, const void* buffer, size_t count) {
+                                               static size_t total = 0;
+                                               std::memset((char*)info + total, 0xFF, count);
+                                               total += count;
+                                               return count;
+                                           },
+                                            [](void* info) { free(info); } };
+    consumer = woc::MakeAutoCF<CGDataConsumerRef>(CGDataConsumerCreate(largeCompareData, &largeCallbacks));
+    written = _CGDataConsumerWriteData(consumer, rawData, 50000);
+    EXPECT_EQ(50000, written);
+    EXPECT_EQ(0xFF, largeCompareData[0]);
+    EXPECT_EQ(0xFF, largeCompareData[49999]);
+    EXPECT_EQ(0, largeCompareData[50000]);
+
+    char smallData[100]{ 0 };
+    CGDataConsumerCallbacks smallCallbacks{ [](void* info, const void* buffer, size_t count) -> size_t {
+                                               memcpy(info, buffer, 50);
+                                               return 50;
+                                           },
+                                            nullptr };
+
+    consumer = woc::MakeAutoCF<CGDataConsumerRef>(CGDataConsumerCreate(smallData, &smallCallbacks));
+    written = _CGDataConsumerWriteData(consumer, rawData, 50000);
+    EXPECT_EQ(50, written);
+    EXPECT_EQ(0, smallData[0]);
+    EXPECT_EQ(1, smallData[1]);
+    EXPECT_EQ(49, smallData[49]);
+    EXPECT_EQ(0, smallData[50]);
+}
+
+TEST(CGDataConsumer, CreateWithCFData) {
+    EXPECT_EQ(nullptr, CGDataConsumerCreateWithCFData(nullptr));
+
+    auto data = woc::MakeAutoCF<CFMutableDataRef>(CFDataCreateMutable(nullptr, 0));
+    auto consumer = woc::MakeAutoCF<CGDataConsumerRef>(CGDataConsumerCreateWithCFData(data));
+    ASSERT_NE(nullptr, consumer);
+    char rawData[100];
+    std::iota(rawData, rawData + 100, 0);
+    size_t written = _CGDataConsumerWriteData(consumer, rawData, 100);
+    EXPECT_EQ(100, written);
+
+    const UInt8* dataPtr = CFDataGetBytePtr(data);
+    EXPECT_EQ(0, std::memcmp(rawData, dataPtr, 100));
+}
+
+TEST(CGDataConsumer, CreateWithURL) {
+    EXPECT_EQ(nullptr, CGDataConsumerCreateWithURL(nullptr));
+
+    auto url = woc::MakeAutoCF<CFURLRef>(CFURLCreateWithFileSystemPath(nullptr, CFSTR("./testfile.txt"), kCFURLWindowsPathStyle, NO));
+    auto consumer = woc::MakeAutoCF<CGDataConsumerRef>(CGDataConsumerCreateWithURL(url));
+    ASSERT_NE(nullptr, consumer);
+
+    const char* dataToWrite = "TEST123";
+    EXPECT_NE(0, _CGDataConsumerWriteData(consumer, dataToWrite, 7));
+    EXPECT_NE(0, _CGDataConsumerWriteData(consumer, dataToWrite, 4));
```

Comment:
You may have to release the consumer here to make utterly certain the buffer has been flushed. It is technically allowable for the I/O system to buffer these two writes and for them to not hit the disk.